XMonad/Layout/PerScreen.hs

```diff
@@ -1,75 +1,119 @@
 
 module XMonad.Layout.PerScreen
     ( -- * Usage
       -- $usage
-      PerScreen,
+      OnScreen,
+      onScreen,
+      onScreens,
+      -- * Deprecated
+      -- $deprecated
+      BW.PerScreen,
       ifWider
     ) where
 
 import XMonad
 import qualified XMonad.StackSet as W
 
-import XMonad.Prelude (fromMaybe)
+import XMonad.Prelude (fromMaybe, fi)
+
+import qualified XMonad.Layout.ByWidth as BW
+
+import Data.List (find)
 
 -- $usage
 -- You can use this module by importing it into your ~\/.xmonad\/xmonad.hs file:
 --
--- > import XMonad.Layout.PerScreen
+-- > import XMonad.Layout.OnScreen
 --
 -- and modifying your layoutHook as follows (for example):
 --
--- > layoutHook = ifWider 1280 (Tall 1 (3/100) (1/2) ||| Full) Full
+-- > layoutHook = onScreen 1 (Tall 1 (3/100) (1/2) ||| Full) Full
 --
 -- Replace any of the layouts with any arbitrarily complicated layout.
--- ifWider can also be used inside other layout combinators.
+-- 'onScreen' can also be used inside other layout combinators, although the
+-- result may be confusing.
+
+-- | Specify a layout to run on a given screen.
+onScreen :: (LayoutClass l1 a, LayoutClass l2 a)
+         => ScreenId -> l1 a -> l2 a -> OnScreen l1 l2 a
+onScreen s = onScreens [s]
+
+-- | Specify a layout to run on a list of screens.
+--   Note that this works by 'ScreenId'. It has a 'Num' instance, so literal
+--   screen numbers will work as expected, but if you use a binding you need
+--   to use the 'S' constructor.
+onScreens :: (LayoutClass l1 a, LayoutClass l2 a)
+          => [ScreenId] -> l1 a -> l2 a -> OnScreen l1 l2 a
+onScreens ss l1 l2 = OnScreen ss l1 l2 False -- @@@ is this right?
+
+data OnScreen l1 l2 a = OnScreen [ScreenId] (l1 a) (l2 a) Bool
+    deriving (Read, Show)
+
+instance (LayoutClass l1 a, LayoutClass l2 a, Show a) => LayoutClass (OnScreen l1 l2) a where
+    runLayout (W.Workspace i p@(OnScreen ss l1 l2 _) ms) r = do
+        -- this sucks.
+        -- You might think we could use the existing screen detail, but we may be
+        -- invoked from 'rescreen' in which case only 'rescreen' and 'windows'
+        -- know the new screen layout. So we need to duplicate the 'rescreen'
+        -- logic (and pray there's no TOCTOU) to get the current screen rects.
+        -- It'd be easier if 'runLayout' were passed the whole 'StackSet' with
+        -- 'windows's idea of the current screen rects, but then it wouldn't know
+        -- which 'Workspace' to work on…
+        which <- withDisplay getCleanedScreenInfo >>= \case
+            [] -> trace "OnScreen: no screens?" >> return False
+            rs -> return $ maybe 0 fst (find (f r) (zip [0..] rs)) `elem` ss
+                where f lr (_,sr) = rect_x lr >= rect_x sr &&
+                                    rect_x lr < rect_x sr + fi (rect_width sr) &&
+                                    rect_y lr >= rect_y sr &&
+                                    rect_y lr < rect_y sr + fi (rect_height sr)
```

Comment:
I agree, this does indeed suck. Maybe we hide it in `XMonad.Prelude.deriveActiveScreen`? It would make the code easier to read

Reply:
Most of the ugliness there is just testing if the (top left corner of the) current screen rectangle is contained in a screen, no? The rest is the `zip` which is a bit deeply embedded to be yanked out.

Reply:
And I'd have a different kind of ugliness if I were building a list of screen rectangles from `ScreenDetail`s anyway.